lib/puppet/configurer.rb

```diff
@@ -207,7 +207,7 @@ def run_internal(options)
       if catalog = prepare_and_retrieve_catalog_from_cache
         options[:catalog] = catalog
         @cached_catalog_status = 'explicitly_requested'
-        @environment = catalog.environment
+        @environment = catalog.environment unless Puppet[:strict_environment_mode]
```

Comment:
Two things here:

1) What should the `environment` field of the report say when `strict_environment_mode` is set? With the way the code is not, If we're using a cached catalog, the report will include the environment of the catalog (along with the fact that the run failed and the error message). Does that seem sane, or should `environment` be empty or the agent specified environment?

2) Unrelated to `strict_environment_mode`, I realized that we probably should log some info saying that we're changing environments when using cached catalogs. We do it in the convergence loop with "Local environment doesn't match blah, restarting agent run with blah.." I'm going to add a second commit to add a notice here.